docs/proposals/TrafficRoute-L4.md

```diff
@@ -0,0 +1,328 @@
+  namespace: demo-app
+spec:
+  rules:
+  - sources:
+    - match:
+        service: mobile.mobile-app.svc:8080
+        region: us-east-1
+        version: 10
+    destinations:
+    - match:
+        # NOTE: only `service` tag can be used here (on `k8s` all TCP connections will have the same Virtual IP as destination => it's not enough info to infer any other destination tags)
+        service: backend.backend-app.svc:8080
+    conf:
+    - weight: 90
+      destination:
+        service: backend
+        region: us-east-1
+        version: 2
+    - weight: 10
+      destination:
+        service: backend.backend-app.svc:8080
+        region: eu-west-3
+        version: 3
+```
+
+NOTE:
+
+1. Notice `match:` syntax (which stems from limitations of `Protobuf`)
+2. `service` tag get generated automatically on `Kubernetes`. The format is `<Service Name>.<Namespace>.svc:<Service Port>`
+3. Read a note above regarding constraints on tags that can be used in `destination`
+
+### Universal
+
+```yaml
+type: TrafficRoute
+name: rule-name
+mesh: default
+rules:
+- sources:
+  - match:
+      service: mobile
+      region: us-east-1
+      version: 10
+  destinations:
+  - match:
+        # NOTE: only `service` tag can be used here (in `universal` all TCP connections will have `127.0.0.1` as destination => it's not enough info to infer any other destination tags)
+      service: backend
+  conf:
+  - weight: 90
+    destination:
+      service: backend
+      region: us-east-1
+      version: 2
+  - weight: 10
+    destination:
+      service: backend
+      version: 3
+```
+
+NOTE:
+
+1. Notice `match:` syntax (which stems from limitations of `Protobuf`)
+2. Read a note above regarding constraints on tags that can be used in `destination`
+
+## Consideration Notes 
+
+* L4 load-balancing in `Envoy` is implemented by `envoy.tcp_proxy` network filter
+* `envoy.tcp_proxy` does support weighted routing to multiple destination clusters, BUT it does NOT support routing to a subset of endpoints within that cluster. E.g.
+  ```
+    destination:
+      service: backend
+      region: us-east-1
+      version: 2
+  ```
+  is a subset of endpoints inside the cluster for `service` "*backend*", but `envoy.tcp_proxy` can only balance to the while cluster `service` "*backend*"
+
+  * possible solutions:
+    1. create an "ad-hoc" cluster that includes only those endpoints that match `destination` (this way we're effectively ignoring subset-based load balancing in Envoy) - `Istio` does that
+       * cons:
+         * "ad-hoc" cluster will have a unique name different from "*backend*". Which has an impact on `StatsD/Prometheus` metrics emitted by Envoy
+         * larger memory footprint (can ignore it for now)
+    2. contribute to `Envoy` to implement subset-based load balancing at L4
+       * cons:
+         * it will take time to merge it upstream and before it will appear in Envoy release
+         * it's strange that `Istio` hasn't implemented it yet
```

Comment:
Can we ask this question on Istio/Envoy Slack? Maybe there are things that we did not take into account.